src/primitives/sinsemilla.rs

```diff
@@ -114,13 +116,11 @@ impl HashDomain {
     fn hash_to_point_inner(&self, msg: impl Iterator<Item = bool>) -> IncompletePoint {
         let padded: Vec<_> = Pad::new(msg).collect();
 
-        let hasher_S = pallas::Point::hash_to_curve(S_PERSONALIZATION);
-        let S = |chunk: &[bool]| hasher_S(&lebs2ip_k(chunk).to_le_bytes());
-
         padded
             .chunks(K)
             .fold(IncompletePoint::from(self.Q), |acc, chunk| {
-                (acc + S(chunk)) + acc
+                let (S_x, S_y) = SINSEMILLA_S[lebs2ip_k(chunk) as usize];
+                acc + pallas::Affine::from_xy(S_x, S_y).unwrap().to_curve() + acc
```

Comment:
This _should_ be fine (either precedence would give the same operation order, but in keeping with the protocol spec I'd prefer the braces be retained. I'd also prefer to construct `S` separately:
```suggestion
                let S_chunk = pallas::Affine::from_xy(S_x, S_y).unwrap().to_curve();
                (acc + S_chunk) + acc
```